File: frontend/lib/validation.ts
export type ValidationError =
  | "InvalidEmailAddress"
  | "EmailTooLong"
  | "PasswordNotAscii"
  | "PasswordTooLong"
  | "PasswordTooShort";

const EMAIL_REGEX = /^[\w\.-]+@[a-zA-Z\d\.-]+\.[a-zA-Z]{2,}$/;

export function validateEmail(email: string): ValidationError[] {
  const errors: ValidationError[] = [];

  if (!EMAIL_REGEX.test(email)) {
    errors.push("InvalidEmailAddress");
  }
  if (email.length > 254) {
    errors.push("EmailTooLong");
  }

  return errors;
}

export function validatePassword(password: string): ValidationError[] {
  const errors: ValidationError[] = [];

  // Check if password contains only ASCII characters
  if (!/^[\x00-\x7F]*$/.test(password)) {
    errors.push("PasswordNotAscii");
  }
  if (password.length > 64) {
    errors.push("PasswordTooLong");
  }
  if (password.length < 8) {
    errors.push("PasswordTooShort");
  }

  return errors;
}

export function validateAuthInput(
  email: string,
  password: string,
): ValidationError[] {
  return [...validateEmail(email), ...validatePassword(password)];
}

export function getErrorMessage(error: ValidationError): string {
  switch (error) {
    case "InvalidEmailAddress":
      return "Please enter a valid email address";
    case "EmailTooLong":
      return "Email address is too long (max 254 characters)";
    case "PasswordNotAscii":
      return "Password must contain only ASCII characters";
    case "PasswordTooLong":
      return "Password is too long (max 64 characters)";
    case "PasswordTooShort":
      return "Password must be at least 8 characters";
    default:
      return "An error occurred";
  }
}

File: ios/tofustash/Auth/Validation.swift
import Foundation

// These helpers define what the auth forms treat as acceptable user input.
// They return explicit validation errors so the UI can show all relevant
// feedback at once instead of failing one field at a time.
//
// `#"..."#` is Swift's raw string literal syntax, which keeps regexes readable.
private let emailRegex = try! NSRegularExpression(
    pattern: #"^[\w\.\-]+@[a-zA-Z\d\.\-]+\.[a-zA-Z]{2,}$"#
)

func validateEmail(_ email: String) -> [ValidationError] {
    var errors: [ValidationError] = []

    // NSRegularExpression is an older Foundation API, so Swift strings must be
    // bridged into `NSRange` before matching.
    let range = NSRange(email.startIndex..., in: email)
    if emailRegex.firstMatch(in: email, range: range) == nil {
        errors.append(.invalidEmailAddress)
    }
    if email.count > 254 {
        errors.append(.emailTooLong)
    }

    return errors
}

func validatePassword(_ password: String) -> [ValidationError] {
    var errors: [ValidationError] = []

    // Passwords are intentionally restricted to ASCII here so backend and client
    // validation behave the same for every user.
    if !password.allSatisfy({ $0.asciiValue != nil }) {
        errors.append(.passwordNotAscii)
    }
    if password.count > 64 {
        errors.append(.passwordTooLong)
    }
    if password.count < 8 {
        errors.append(.passwordTooShort)
    }

    return errors
}

func validateAuthInput(email: String, password: String) -> [ValidationError] {
    // Concatenate the two error arrays so the UI can show both field problems
    // from a single submit attempt.
    return validateEmail(email) + validatePassword(password)
}
